### PR TITLE
Add CLASS or RUNTIME limitation to annotation processors

### LINKED ISSUE #19366: Wrong code gets generated when using MapStruct decorators with incremental compilation in Gradle 7.3
I've ran into a problem with using MapStruct's annotation processor with incremental compilation after upgrading to Gradle 7.3, I have also created a MapStruct issue here: https://github.com/mapstruct/mapstruct/issues/2682

### Expected Behavior

Expected the generated `PersonaMapperImpl.java` after incremental compilation to be:

```java
@Generated(
    value = "org.mapstruct.ap.MappingProcessor",
    date = "2021-12-08T14:16:55+1300",
    comments = "version: 1.5.0.Beta1, compiler: IncrementalProcessingEnvironment from gradle-language-java-7.3.1.jar, environment: Java 17 (Private Build)"
)
@Component
@Primary
class PersonMapperImpl extends PersonMapperDecorator {
}
```

### Current Behavior

Actual content of `PersonaMapperImpl.java`:

```java
@Generated(
    value = "org.mapstruct.ap.MappingProcessor",
    date = "2021-12-08T14:14:11+1300",
    comments = "version: 1.5.0.Beta1, compiler: IncrementalProcessingEnvironment from gradle-language-java-7.3.1.jar, environment: Java 17 (Private Build)"
)
@Component
class PersonMapperImpl implements PersonMapper {

    @Override
    public PersonDto personToPersonDto(Person person) {
        if ( person == null ) {
            return null;
        }
        PersonDto personDto = new PersonDto();
        personDto.name = person.name;
        return personDto;
    }
}
```

### Context

Used to work as expected in Gradle 7.2 but no longer works in 7.3, I have to force a full compilation to get the expected code to be generated, which is not practical for a large project.

### Steps to Reproduce 

1. Check out the demo project for this issue: https://github.com/laech/mapstruct-spring-gradle-incremental-issue
2. `./gradlew compileJava` (`PersonaMapperImpl.java` has the correct content at this point in time)
3. Then make a change in `PersonMapperDecorator.java`, such as adding a random comment.
4. `./gradlew compileJava` (this will run an incremental build,  `PersonaMapperImpl.java` now has the wrong content)

### Your Environment

Ubuntu 21.10
JDK 17

### PR DESCRIPTION
**Pull request by asodja**
Fixes #19366

### CODE REVIEW

**Comment by asodja** [comment]
@oehme could you take a look?

File: subprojects/docs/src/docs/userguide/jvm/java_plugin.adoc [hunk 1]
```diff
@@ -546,6 +546,8 @@ include::{snippetsPath}/java/incrementalAnnotationProcessing/groovy/processor/sr
 
 Both categories have the following limitations:
 
+* They have to read `CLASS` or `RUNTIME` retention annotations only.
+    Reading `SOURCE` retention annotations is possible for "isolating" processors on their own. But it will produce incorrect results when a different "aggregating" processor process the same types as the "isolating" processor, so it should not be used.
```
Comment (oehme) [inline subprojects/docs/src/docs/userguide/jvm/java_plugin.adoc:550]: This sentence just weakens the point above and could lead to confusion, I'd remove it completely.

File: subprojects/docs/src/docs/userguide/jvm/java_plugin.adoc [hunk 2]
```diff
@@ -596,9 +598,8 @@ For instance, a `ServiceRegistryProcessor` could create a single `ServiceRegistr
 include::{snippetsPath}/java/incrementalAnnotationProcessing/groovy/processor/src/main/java/org/gradle/ServiceRegistryProcessor.java[tag=aggregating-annotation-processor]
 ----
 
-"Aggregating" processors have the following limitations:
+"Aggregating" processors have the following additional limitation:
 
-* They can only read `CLASS` or `RUNTIME` retention annotations
 * They can only read parameter names if the user passes the `-parameters` compiler argument.
```
Comment (oehme) [inline subprojects/docs/src/docs/userguide/jvm/java_plugin.adoc:603]: This also applies to all processors, for the same reason as the class/runtime retention annotations.

**Review by oehme — CHANGES_REQUESTED** [review]

File: subprojects/docs/src/docs/userguide/jvm/java_plugin.adoc (hunk 1 above)
Comment (asodja) [reply subprojects/docs/src/docs/userguide/jvm/java_plugin.adoc]: Removed explanations that might confuse users.

File: subprojects/docs/src/docs/userguide/jvm/java_plugin.adoc (hunk 2 above)
Comment (asodja) [reply subprojects/docs/src/docs/userguide/jvm/java_plugin.adoc]: Moved to `all` section

**Review by oehme — APPROVED** [review]

**Review by wolfs — APPROVED** [review]
LGTM!

**Comment by asodja** [comment]
@bot-gradle test and merge

**Pull request merged**